tests/test_otel.py

```diff
@@ -0,0 +1,367 @@
+        )
+    return resp
+
+
+@pytest.fixture
+def batches(trace_service):
+    for _ in range(20):
+        if len(trace_service.spans):
+            break
+        time.sleep(0.1)
+    else:
+        assert len(trace_service.spans) > 0, "No spans received"
+    yield trace_service.spans
+    trace_service.spans.clear()
+
+
+@pytest.fixture
+def span(batches):
+    assert len(batches) == 1, "Expect 1 batch"
+    assert len(batches[0][0].scope_spans[0].spans) == 1, "Expect 1 span"
+    return batches[0][0].scope_spans[0].spans[0]
+
+
+@pytest.fixture
+def response(logger, http_ver, scheme, path, headers):
+    return get_response(logger, http_ver, scheme, path, headers)
+
+
+@pytest.fixture
+def nresponses(logger, http_ver, scheme, path, headers, n):
+    out = []
+    for _ in range(n):
+        out.append(get_response(logger, http_ver, scheme, path, headers))
+    return out
+
+
+@pytest.mark.usefixtures("trace_service", "otelcol", "nginx")
+@pytest.mark.parametrize(
+    ("http_ver", "scheme"),
+    [
+        (0, "http"),
+        (1, "http"),
+        (2, "https"),
+        (3, "https"),
+    ],
+    ids=["http 0.9", "http 1.1", "http 2.0 ssl", "http 3.0 quic"],
+)
+@pytest.mark.parametrize(
+    ("path", "headers", "text", "status"),
+    [
+        ("/", None, "TRACE-ON", 200),
+        ("/404", None, "404 Not Found", 404),
+    ],
+    ids=["ok", "error"],
+)
+def test_response_and_span_attributes(
+    http_ver, scheme, path, text, status, response, span
+):
+    assert text in (response.text if http_ver else response)
+    if http_ver:
+        assert response.status_code == status
+
+    assert span.name == ("default_location" if path == "/" else "/404")
+
+    # Default span attributes
+    assert get_attr(span, "http.method", "string_value") == "GET"
+    assert get_attr(span, "http.target", "string_value") == path
+    assert get_attr(span, "http.route", "string_value") == path
+    assert get_attr(span, "http.scheme", "string_value") == scheme
+    assert get_attr(span, "http.flavor", "string_value") == (
+        [None, "1.1", "2.0", "3.0"][http_ver]
+    )
+    assert get_attr(span, "http.user_agent", "string_value") == (
+        ([None] + [f"niquests/{niquests.__version__}"] * 3)[http_ver]
+    )
+    assert get_attr(span, "http.request_content_length", "int_value") == 0
+    assert get_attr(span, "http.response_content_length", "int_value") == (
+        len(response.text if http_ver else response)
+    )
+    assert get_attr(span, "http.status_code", "int_value") == status
+    assert get_attr(span, "net.host.name", "string_value") == "localhost"
+    assert get_attr(span, "net.host.port", "int_value") == (
+        [18080, 18080, 18443, 18443][http_ver]
+    )
+    assert get_attr(span, "net.sock.peer.addr", "string_value") == "127.0.0.1"
+    assert get_attr(span, "net.sock.peer.port", "int_value") in range(
+        1024, 65536
+    )
+
+    # Custom span attributes
```

Comment:
These might deserve their own test without any "parametrize".

Reply:
The "should not"/"absent" case doesn't make any sense - you are testing that module won't add 3 specific attributes.

Reply:
I test that module add custom metrics when it's expected and doesn't where it shouldn't.

Reply:
Removed unexpected var test.